core/backend/src/Entity.ts

```diff
@@ -64,6 +64,11 @@ export class Entity {
    */
   public toJSON(): EntityProps {
     const val: any = {};
+    if (!this.iModel.isOpen) {
+      //! If IModel associated with EntityProps is closed then required metadata cannot be accessed and this operation will fail.
+      throw new IModelError(IModelStatus.NotOpen ,"IModel associated with entity is not open.");
```

Comment:
@pmconne I have added an exception here on `Entity.toJSON()` will surely fail if iModel is closed. I updated two tests and called `toJson()` before the respective iModel was closed.

Reply:
I don't understand why the error isn't thrown where we attempt to access the iModel (e.g. in `nativeDb.getECClassMetadata`). That's where this error should be thrown.

Reply:
That's because nativeDb is undefined in closed iModelDb. That is where the error was thrown that `property nativeDb is undefined`.

Reply:
Though i can put it in following
```
  public get classMetaDataRegistry(): MetaDataRegistry {
    if (this._classMetaDataRegistry === undefined)
      this._classMetaDataRegistry = new MetaDataRegistry();

    return this._classMetaDataRegistry;
  }
  ```

Reply:
Should probably be throwing in `IModelDb.nativeDb` getter instead ...

> `public get nativeDb(): IModelJsNative.DgnDb { return this._nativeDb!; } // eslint-disable-line @typescript-eslint/no-non-null-assertion`

Reply:
> no, nativeDb is never undefined, It is set in the constructor.

See last line of `IModelDb.close`.
```
  /** Close this IModel, if it is currently open. */
  public close(): void {
    if (!this.isOpen)
      return; // don't continue if already closed

    this.beforeClose();
    IModelDb._openDbs.delete(this._fileKey);
    this._workspace?.close();
    this.locks.close();
    this._locks = undefined;
    this._codeService?.close();
    this._codeService = undefined;
    this.nativeDb.closeIModel();
    this._nativeDb = undefined; // the underlying nativeDb has been freed by closeIModel
  }
```

Reply:
After the iModel is closed it is too late to check if the iModel is open in `nativeDb.anything` because `nativeDb` will be `undefined`.
The `nativeDb` getter should be throwing if `_nativeDb` is `undefined`.
Assertions - including `!` - are for things we **know** must always be true, not for things we **hope** are true.

Reply:
i see that we set the _nativeDb member to undefined after you close. I guess that's so you'll get this error. Dunno. Probably should remove that.

Reply:
@khanaffan what are you doing when you hit this error? Do you somehow have an open iModel, close it, and then try to use it after that? Why? Are you expecting to close/open the nativeDb? If so, don't use `iModel.close`, use `nativeDb.closeIModel`.

I don't mind if you remove the `nativeDb=undefined` in `iModelDb.close`. That's only there so you *will* get an error if you attempt to use your iModel after it's closed because we don't have any way to re-open a closed iModel.

Reply:
i removed the getter and setting the member to undefined in close. You'll now get different errors, later. But it does make it possible to close and reopen if we ever have a need for that.

Reply:
please remove this change to toJSON.

Reply:
There are two unit test that does that. It grab a entity and close the connection and then call toJSON() on it. toJSON() at some point need meta data which is cleared when connection is closed. My only change is to clear the meta data on iModelDb.clearCache() after schema import so schema changes are visiable.

Reply:
I am removing it.

Reply:
If the tests are doing that, they will get *some* exception, because what they're doing is illegal. The exception should come from 'getECClassMetadata` indicating that the iModel is closed. That's the necessary fix here.